### PR TITLE
Refactor prep_segments in SVS

### PR DESCRIPTION
**Pull request by jerryuhoo**
@ftshijt @A-Quarter-Mile 
I plan to use only one file for preprocessing labels and XML files, instead of using `local/prep_segments.py` and `local/prep_segments_from_xml.py` in every dataset, which will be placed under `pyscripts/utils`. 
To fix dataset errors, we can use several functions (`replace_lyrics`, `replace_labels`, `skip_labels`, `add_missing_phoneme`, `add_pause`) to handle specific errors. We can use a dictionary `error_correction` to store error handling functions in each dataset.

In order to use this new function, previous scripts for datasets that use XML files should be changed.
1. in local/data.sh, we should use `pyscripts/utils/prep_segments.py` and add parameters such as `--dataset natsume` and `--input_type hts/xml`. Please see Natsume example.
2. All error-handling code for each dataset should be added to this file.
3. After that, delete `local/prep_segments.py`, `local/prep_segments_from_xml.py` and `pyscripts/utils/prep_segments_from_xml.py`

### CODE REVIEW

File: egs2/namine_ritsu_utagoe_db/svs1/local/data.sh [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+#         mv ${src_data}/label.tmp ${src_data}/label
+#         local/prep_segments_from_xml.py --silence P --silence B ${src_data}
+#         mv ${src_data}/text.tmp ${src_data}/text
+#         mv ${src_data}/segments_from_xml.tmp ${src_data}/segments_from_xml
+#         mv ${src_data}/score.scp.tmp ${src_data}/score.scp
+#         awk '{printf("%s namine\n", $1);}' < ${src_data}/segments > ${src_data}/utt2spk
+#         utils/utt2spk_to_spk2utt.pl < ${src_data}/utt2spk > ${src_data}/spk2utt
+#         utils/fix_data_dir.sh --utt_extra_files "label score.scp" ${src_data}
+#     done
+# fi
```
Comment (ftshijt) [inline egs2/namine_ritsu_utagoe_db/svs1/local/data.sh:67]: Let's safely remove it if not used.

File: egs2/namine_ritsu_utagoe_db/svs1/local/dataset_split.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        wavscp.write(
+            "{} {}\n".format(
+                utt_id, os.path.join(wav_dump, "{}_bits16.wav".format(folder))
+            )
+        )
+        utt2spk.write("{} {}\n".format(utt_id, UTT_PREFIX))
+        label_info, text_info = process_text_info(
+            os.path.join(src_data, folder, "{}.lab".format(folder))
+        )
+        # text_scp.write("{} {}\n".format(utt_id, text_info))
```
Comment (ftshijt) [inline egs2/namine_ritsu_utagoe_db/svs1/local/dataset_split.py:86]: Why do you remove the write of text_scp and midiscp?

**Review by ftshijt — COMMENTED** [review]
super thanks! it would be great if we can create a better doc in egs/TEMPLATE/svs1 to instruct people how to use the tool

**Comment by A-Quarter-Mile** [comment]
The modification functions in new prep_segment look great. But it would be too long if we put all datasets in it.

File: egs2/namine_ritsu_utagoe_db/svs1/local/dataset_split.py (hunk 2 above)
Comment (jerryuhoo) [reply egs2/namine_ritsu_utagoe_db/svs1/local/dataset_split.py:86]: > Why do you remove the write of text_scp and midiscp?

I think they are not used in this dataset. I find that ofuton also doesn't use text_scp and midiscp, so I removed them. Should I keep them?

**Comment by jerryuhoo** [comment]
> 

I see. I think it would be a good idea to generate a dict in check_align.py, then I can use the dict in prep_segments.py. In that case, all the errors will be auto-fixed, and we don't have to put those dicts inside the code.

**Comment by ftshijt** [comment]
Looks good to me. Could you also add the related docs in template/svs1/readme ? I'm still not pretty sure about the potential steps we need to do with the regarding updates.

File: egs2/TEMPLATE/svs1/README.md [hunk 3]
```diff
@@ -565,6 +565,8 @@ Then, we transfer the raw data into `score.json`, where situations can be catego
 
 - If the phoneme annotation are misaligned with notes in time domain, align phonemes (from `label`) and note-lyric pairs (from `musicXML`) through g2p. (eg. [Ofuton](https://github.com/espnet/espnet/tree/master/egs2/ofuton_p_utagoe_db/svs1))
 
+- For some datasets such as natsume or namine, there is a method to automatically repair missing silences in the dataset. When you run stage 1 and encounter errors such as "Lyrics are longer than phones" or "Phones are longer than lyrics", the code that needs to be fixed will be automatically printed out. You only need to copy this code into the `get_error_dict` method in `egs2/TEMPLATE/svs1/pyscripts/utils/prep_segments.py`. Depending on the suggested input_type, copy it into the hts or xml's error_dict respectively.
```
Comment (ftshijt) [inline egs2/TEMPLATE/svs1/README.md:568]:
```suggestion
- We also offer some automatic fixes for missing silences in the dataset. During the stage1, when you encounter errors such as "Lyrics are longer than phones" or "Phones are longer than lyrics", the scripts will auto-generated the fixing code. You may need to put the code into the `get_error_dict` method in `egs2/TEMPLATE/svs1/pyscripts/utils/prep_segments.py`. Noted that depending on the suggested input_type, you may want to copy it into either the `hts` or `xml`'s error_dict. (For more information, please check [namine](https://github.com/espnet/espnet/tree/master/egs2/namine_ritsu_utagoe_db/svs1) or [natsume](https://github.com/espnet/espnet/tree/master/egs2/natsume/svs1)
```

File: egs2/namine_ritsu_utagoe_db/svs1/local/dataset_split.py (hunk 2 above)
Comment (ftshijt) [reply egs2/namine_ritsu_utagoe_db/svs1/local/dataset_split.py:90]: If it is not used, let's remove them.

**Comment by ftshijt** [comment]
LGTM! @A-Quarter-Mile could you double check it?

**Comment by A-Quarter-Mile** [comment]
I still suggest removing natsume from new prep_segments and keeping the one in local/. There are some special designs for it (mid_reader and align_lyric_notes), which are not applicable for other datasets. Besides, the modifications for natsume in new prep_segments are based on manual inspection, we can remove them safely.

**Comment by ftshijt** [comment]
LGTM! @A-Quarter-Mile could you double check this update? After your confirmation and CI cleared, I will merge the PR.

**Comment by ftshijt** [comment]
Many thanks! Code merged.

**Pull request merged**